### ISSUE TITLE
[v6 beta 9] Section and line background colors

### ISSUE THREAD

**Issue by iv-gha** [thebigg73/OpenSongTablet#198]
Hello Gareth,

1. Section background color should fill the width of the song column (v5 behaviour). 
2. Line background color of comment lines should fill the the width of the song column,

Where a section is smaller than the column width (font size restricted) a lines background color does not fill the column width..  v6 does not yet generate full width lines, so background color is also not full width.  A change to the views after they are measured and before final layout may be possible - this to make the width match the parent so that color fills the column.

Please consider.

**Comment by thebigg73** [comment]
Hi Ian,

This was intentional from me due to the way the views are scaled.  In v5, the lines were iterated through adding spaces to make them the same size.  They were then drawn several times and measured.  This was due to individual lines having their background colours set.  Now the entire section colour is set as one item, so this get set to the width of the widest line.
This was more efficient for coding in my opinion and this, along with drawing once, measuring and scaling, rather than redrawing with a larger font size accounts for the majority of the processing speed.  Making the individual section views fill the width would break the measuring of them required for scaling.

I will look at adding a frame layout behind each section of final render to the screen that fits the width. and places the scaled section inside this.

**Comment by iv-gha** [comment]
Thanks for considering.  The most obvious 'odd' look is for stage mode.
Varying section widths being evident when limiting font size.

A section may contain comment too, with this having background of the
right color but the width of the section.

The ideal would be that the line background colors are full screen width -
recognising that this will need a tweak to v6 views for final render. We do
not want v6 Stage render to look odd!

**Comment by thebigg73** [comment]
Had a go at this for beta 10, however, points to note:
Performance mode:  The background colours will now match the width of the song's widest view, but not the screen width. 
 This is required to allow correct measuring and scaling of individual sections.  It also lets users see if they have whitespace entered into at the end of a song line (I had this for lots of songs written a while back).
Stage mode:  Since the comment lines now pull in a header (from the previous section), I can't get a way of getting the correct background colour for the entire line/section.  The view is coloured, but uses the colour of the section, which doesn't match the comment background.  I'm struggling to find a way around this just now.

**Comment by iv-gha** [comment]
A thought... One method could be to store views twice during layout - one for size, one for final layout. This may allow the adding of an additional view to a line for the final layout version to be set to size to the end if the line?  Should not slow things down?

**Comment by thebigg73** [comment]
Unfortunately views cannot be copied - they would have to be created again by a second pass through the logic.   You'll see in the code I have to display measure them involves removing them from parents before reusing them.  Also I'd worry about the memory implications given the number of views in a song (each chord/lyric line has multiple views as they are split into table cells for alignment).

**Comment by thebigg73** [comment]
As you'll see in beta 10, it is only the header for comment lines that doesn't match in stage mode.  I think it looks fine.  Attached screenshots for performance and stage modes:
![image](https://user-images.githubusercontent.com/7831274/224482669-895875a3-b542-4b25-9ac0-f07bf7762af5.png)
![image](https://user-images.githubusercontent.com/7831274/224482693-ef24717e-e179-4e53-be25-0f985f803987.png)

**Comment by iv-gha** [comment]
That looks good.  The comment is a section because of double new line?  I am thinking this is a section split issue.  I will take a look.  Great!

**Comment by iv-gha** [comment]
Hi, first look at 10 suggests sizing has been badly disturbed in Stage mode.  In general I run with a font size of say 24 to get near consistent char size (small lines from double column Songselect downloads help).  Now get v.small font sizes / unusable.  I suspect you tried with larger font sizes which masked the problem?

Can you copy me the raw lines of the 'chorus with the problem with comment lines' seen earlier - thanks.

**Comment by thebigg73** [comment]
Hi Ian,
I haven't changed the font sizes.  They are always drawn to the screen and measured at 14sp.  The app should then scale the view (not the font) to the correct size.  This is the same for all modes.  Can you show a screen shot of what you are seeing?  The screenshot I posted above is what I see in Stage Mode

**Comment by iv-gha** [comment]
![Screenshot_20230312-075850](https://user-images.githubusercontent.com/68599243/224531973-03180f7c-d877-4c06-bdcb-3531d48a7de1.png)
So it is the font size in scaling - set to 26 in this example. Cheers.

**Comment by thebigg73** [comment]
Yes, the scaling works out the maximum scale value to fit the width/height constraints. e.g:
`max_view_scaling = (float)width available/(float) view width `

It then works out the maximum scaling allowed based on your max preferred font size e.g:
`max_font_scaling = (float) maxfont / 14f`

It then uses the minimum of the two
`scaling = Math.min(max_view_scaling, max_font_scaling)`

For stage mode, it also checks the scaling for the y value doesn't exceed the user percentage set.

**Comment by thebigg73** [comment]
You are correct about the double line split

```
[C]
.          A         E     B    C#m
 Bless the Lord O my soul, O my soul
.A           E    Bsus4  B
 Worship His holy name
.          A       C#m   A B  C#m
 Sing like never before, O my soul
.     A            B    A  E
 I'll worship Your holy name
 
;Chorus sometimes
 
[V2]...
```

Removing the extra line above stops the section splitting

**Comment by iv-gha** [comment]
Thanks Gareth.  I think this is the previous wrong size sections thing. Opening option menu and close (new open) gives...
![Screenshot_20230312-084131](https://user-images.githubusercontent.com/68599243/224533928-e506def3-777c-4dad-aa85-446ee3ceb62a.png)

So swipe to song bad.  Fresh open good.

Cheers

**Comment by iv-gha** [comment]
Max autoscale font size is set to 26 - this is done deliberately to get most sections using the same font size, a more 'normal' looking layout.  Rather than using a higher max size which results in wide variation in font size depending on section content. A large max size, perhaps your choice, may 'hide' the problem.

**Comment by thebigg73** [comment]
I see small sections if I lower the max preferred font size as well, but this is the choice of the user?  Do you feel we should be ignoring this in Stage Mode and making the text bigger?  Currently the size is the same for Performance and Stage mode

**Comment by thebigg73** [comment]
![setting1](https://user-images.githubusercontent.com/7831274/224541131-dcc569f8-e9cf-4b08-8c51-084d12cb9f13.png)
![stage1](https://user-images.githubusercontent.com/7831274/224541142-e4497c28-a7b6-404c-bac7-af4398e1a899.png)

**Comment by thebigg73** [comment]
![setting2](https://user-images.githubusercontent.com/7831274/224541152-efada0e5-0863-4282-acff-83e6dd97bd93.png)
![stage2](https://user-images.githubusercontent.com/7831274/224541156-47a07d67-bbbf-4425-93b3-707085529b61.png)

**Comment by thebigg73** [comment]
![setting3](https://user-images.githubusercontent.com/7831274/224541164-48a23146-89c9-4e4e-aa6c-02706be7c988.png)
![stage3](https://user-images.githubusercontent.com/7831274/224541169-fbc508cf-0c24-4244-9eac-cbab8c1a15b1.png)

**Comment by thebigg73** [comment]
However, I can see that the size in Stage Mode is smaller than Performance mode - I will investigate!

**Comment by thebigg73** [comment]
Arrgghh.  Figured it out.  The font size used for view creation was 14sp in v5, but I actually changed it to 8sp in v6 as I was previewing on screen during development.  The calculation in StageModeAdapter uses a hard coded 14f for the font size, but it should've been 8!  I've now coded it to pull the size from the ProcessSong value so I don't forget in future!

I'm struggling to upload changes just now as my home internet is down.  I'll push once it is back up and running.

**Comment by iv-gha** [comment]
Hi Gareth, glad you have found root cause.  There must be a reset in some way on use of the options menu - as it is (was) then correctly displayed.  I would compare v6 behaviour with v5 when odd render / sizing issues appear - v5 is proven stable.  As a regular v5 Stage mode user with second screen projection of lyrics -  it stands out to me when either are off in v6.  Cheers  Ian.

**Comment by iv-gha** [comment]
Hello Gareth,
Can you test correct section widths by...
Set a small max font size - sufficient to not quite fill the width.
Open a song in stage mode.
Swipe off and return to the song.
Rotate the screen
Rotate back

All of these should render properly.  Does this work for you?

**Comment by iv-gha** [comment]
Hi,
Sync of the project to my repository...
I am not able to compile as 'import com.garethevans.church.opensongtablet.customviews.GlideApp' is not working .  Is this something that needs to be uploaded to your repository?

**Comment by iv-gha** [comment]
Also

`myView.recyclerView.smoothScrollTo(requireContext(), recyclerLayoutManager, 0);`

 is not found?

**Comment by iv-gha** [comment]
I have shelved all of my work and compiled beta 10 rebased from your MaterialApp.  I think I am compiling what you have.

 Please confirm that for you that Stage Mode is working... that is...
1.  With a small Max scaling font size, say 20
2. Swiping to a song shows correctly sized sections (none too small).
3. Return to the same song using swipe shows correctly sized sections.
4. Rotations are stable with song correctly displaying with multiple rotations.
5. The song is sizing the same when returned to after use of the options menu.

2-4 are bad for me in Stage mode.  Sections become small, the same song can show different small sections on rotation and on return when swiped.  It quickly becomes unstable.  Weird.

It might be something at my end... that is why I have re-based from your repository.  At the moment I am trying to get exactly the same as you!  Doh!  Let me know.

**Comment by thebigg73** [comment]
Tried what you suggested and saw the same issue. 
Found an issue that I had changed various widths trying to get the background colour to draw correctly to the full width
Had this line in the stagemodeadapter that should've been removed.  This resized the actual views to the width of the screen meaning the scaling reset to 1f, meaning the font size ended up being 8sp.  I hope this is the correct fix.  Tested for a while and hopefully sorts it.  Will push now
